src/ndb.c

```diff
@@ -0,0 +1,369 @@
+ * and if it doesn't creates empty file automatically.
+ */
+enum ndb_op_code ndb_open_file(ndb_file_t *file, char *version)
+{
+  char ver[MAX_NDB_FILE_VERSION_LEN];
+
+  if (!do_writes_to_flash) {
+    return NDB_ERR_NONE;
+  }
+
+  file->fh = cfs_open(file->name, CFS_READ + CFS_WRITE);
+  if (file->fh < 0) {
+    cfs_coffee_reserve(file->name, file->expected_size);
+    cfs_coffee_configure_log(file->name, 256 * 2, 256);
+    file->fh = cfs_open(file->name, CFS_READ + CFS_WRITE);
+    if (file->fh < 0) {
+      return NDB_ERR_FILE_IO;
+    }
+  }
+
+  if (cfs_read(file->fh, ver, sizeof(ndb_file_version))
+      == sizeof(ndb_file_version)) {
+    if (memcmp(&ver, version, sizeof(ndb_file_version)) == 0) {
+      return NDB_ERR_NONE;
+    } else {
+      cfs_close(file->fh);
+      cfs_remove(file->name);
+      return ndb_open_file(file, version);
+    }
+  }
+
+  if (cfs_seek(file->fh, 0, CFS_SEEK_SET) != 0) {
+    cfs_close(file->fh);
+    file->fh = -1;
+    return NDB_ERR_FILE_IO;
+  }
+
+  if (cfs_write(file->fh, version, sizeof(ndb_file_version))
+      != sizeof(ndb_file_version)) {
+    cfs_close(file->fh);
+    file->fh = -1;
+    return NDB_ERR_FILE_IO;
+  }
+
+  return NDB_ERR_NONE;
+}
+
+enum ndb_op_code ndb_write_file_data(ndb_file_t *f, cfs_offset_t o, void *b,
+                                     size_t l)
+{
+  if (!do_writes_to_flash) {
+    return NDB_ERR_NONE;
+  }
+
+  int offset = sizeof(ndb_file_version) + o;
+  int ret = cfs_seek(f->fh, offset, CFS_SEEK_SET);
+
+  if (ret == offset) {
+    if (cfs_write(f->fh, b, l) == (int) l) {
+      return NDB_ERR_NONE;
+    }
+  }
+  return NDB_ERR_FILE_IO;
+}
+
+/**
+ * Reading from NDB file.
+ * This function is to be called only during initialization of NDB.
+ * It is intended to read all information elements collection from
+ * file into memory.
+ */
+enum ndb_op_code ndb_read(ndb_file_t *file, void *first_element, size_t size)
+{
+  if (!do_writes_to_flash) {
+    memset(first_element, 0, size);
+    return NDB_ERR_NONE;
+  }
+
+  if (file->fh < 0) {
+    return NDB_ERR_FILE_IO;
+  }
+
+  if (cfs_seek(file->fh, sizeof(ndb_file_version), CFS_SEEK_SET)
+      == (cfs_offset_t) -1) {
+    cfs_close(file->fh);
+    file->fh = -1;
+    return NDB_ERR_FILE_IO;
+  }
+  return
+      cfs_read(file->fh, first_element, size) == (int) size ?
+          NDB_ERR_NONE : NDB_ERR_FILE_IO;
+}
+
+void ndb_lock()
+{
+  chMtxLock(&data_access);
+}
+
+void ndb_unlock()
+{
+  chMtxUnlock(&data_access);
+}
+
+void ndb_retrieve(void* out, void* cached, size_t size)
+{
+  ndb_lock();
+  memcpy(out, cached, size);
+  ndb_unlock();
+}
+
+enum ndb_op_code ndb_update(void* cached, void* new, size_t size,
+                            enum ndb_data_source src,
+                            ndb_element_metadata_t *md, u8 and_state_mask,
+                            u8 or_state_mask)
+{
+  ndb_lock();
+  if (memcmp(new, cached, size) == 0) {
+    ndb_unlock();
```

Comment:
Not sure. Though I'm a little worried about excessive writes to flash. I haven't done any calculations on lifetime.